mlflow/models/signature.py

```diff
@@ -0,0 +1,449 @@
+        if col.dtype.itemsize < 4 or col.dtype.kind == "i" and col.dtype.itemsize == 4:
+            return DataType.integer
+        elif col.dtype.itemsize < 8 or col.dtype.kind == "i" and col.dtype.itemsize == 8:
+            return DataType.long
+    elif col.dtype.kind == "f":
+        if col.dtype.itemsize <= 4:
+            return DataType.float
+        elif col.dtype.itemsize <= 8:
+            return DataType.double
+    elif col.dtype.kind == "U":
+        return DataType.string
+    elif col.dtype.kind == "S":
+        return DataType.binary
+    elif col.dtype.kind == "O":
+        first_elem = col[0]
+        if isinstance(first_elem, bytearray) or isinstance(first_elem, bytes) and all(
+                [isinstance(x, bytearray) or isinstance(x, bytes) for x in col]):
+            return DataType.binary
+        elif isinstance(first_elem, str) and all([isinstance(x, str) for x in col]):
+            return DataType.string
+        elif isinstance(first_elem, int) and all([isinstance(x, int) for x in col]):
+            return DataType.long
+        elif isinstance(first_elem, float) and all([isinstance(x, float) for x in col]):
+            return DataType.double
+        else:
+            raise MlflowException("unsupported element type {} ".format(type(first_elem)))
+    raise MlflowException("Unsupported numpy data type '{0}', kind '{1}'".format(
+        col.dtype, col.dtype.kind))
+
+
+def _map_spark_type(x) -> DataType:
+    import pyspark.sql.types
+    if isinstance(x, pyspark.sql.types.NumericType):
+        if isinstance(x, pyspark.sql.types.IntegralType):
+            if isinstance(x, pyspark.sql.types.LongType):
+                return DataType.long
+            else:
+                return DataType.integer
+        elif isinstance(x, pyspark.sql.types.FloatType):
+            return DataType.float
+        elif isinstance(x, pyspark.sql.types.DoubleType):
+            return DataType.double
+    elif isinstance(x, pyspark.sql.types.BooleanType):
+        return DataType.boolean
+    elif isinstance(x, pyspark.sql.types.StringType):
+        return DataType.string
+    elif isinstance(x, pyspark.sql.types.BinaryType):
+        return DataType.binary
+    else:
+        raise Exception("Unsupported Spark Type '{}', MLflow schema is only supported for scalar "
+                        "Spark types.".format(type(x)))
+
+
+def _is_spark_df(x) -> bool:
+    try:
+        return isinstance(x, importlib.import_module("pyspark.sql.dataframe").DataFrame)
+    except ImportError:
+        return False
+
+
+def _infer_schema(data: MlflowModelDataset) -> Schema:
+    if hasattr(data, "__len__") and len(data) == 0:
+        return Schema([])
+    if isinstance(data, dict):
+        res = []
+        for col in data.keys():
+            ary = data[col]
+            if not isinstance(ary, np.ndarray):
+                raise TypeError("Data in the dictionary must be of type numpy.ndarray")
```

Comment:
Why is one of these a TypeError and the other an MlflowException?

Reply:
The other is the "must me 1 dimensional" exception? Technically that one is not a TypError as it is the right type (numpy array) but wrong value.